### PR TITLE
Add amenity=clock icon

### LINKED ISSUE #3070: Add rendering for amenity=clock
https://wiki.openstreetmap.org/wiki/Tag:amenity=clock
Quite popular tag, it has 11,8k uses.

Pole clocks are sometimes used as meeting points; sundials are tourist attraction, but both aren't rendered so they are unvisible on a map.

Icon proposal:
amenity clock

### PR DESCRIPTION
**Pull request by james2432**
Fixes #3070 
See comment below due to rework
~From https://www.openstreetmap.org/#map=19/45.01689/-75.64647 (node)~
~Z18:~
image
~Z19:~
image

### CODE REVIEW

**Comment by HolgerJeromin** [comment]
We had this idea in the past:

> I think all allowed clocks should be rendered on zoom 19, but only clocks with visibility=area should be on zoom 18.

**Comment by Tomasz-W** [comment]
Please include in code:
* clock areas (rendered as icon only)
* zoom differentation from https://github.com/gravitystorm/openstreetmap-carto/pull/3135#issuecomment-374839092

**Comment by lakedistrictOSM** [comment]
Will places tagged with both `amenity=clock` and `historic=monument`
(like https://www.openstreetmap.org/node/456474545) get rendered with the clock icon or the monument icon?

**Comment by Tomasz-W** [comment]
> Will places tagged with both amenity=clock and historic=monument
> (like https://www.openstreetmap.org/node/456474545) get rendered with the clock icon or the monument icon?

I've checked this combination with Overpass Turbo and Google Earth, most of examples doesn't have any other function than being a clock, so `amenity=clock` tag should have a pririty there IMO.

**Comment by james2432** [comment]
From: https://www.openstreetmap.org/#map=19/35.78917/-78.82883

image
Reworked to clock zoom 18 and needs visibility=area

**Comment by Tomasz-W** [comment]
> You can’t see in google earth if something has a commemorative function, or maybe you can explain how?

street view :)

**Comment by dieterdreist** [comment]
sent from a phone

> On 22. Mar 2018, at 07:21, Tomasz Wójcik <notifications@github.com> wrote:
> 
> You can’t see in google earth if something has a commemorative function, or maybe you can explain how?
> 
> street view :)
> 


neither, I’d say it requires local knowledge

**Comment by james2432** [comment]
@Tomasz-W I understand now, will tweak :) :+1:

**Comment by kocio-pl** [comment]
It's important to test how does it look with churches/towers and stations - people may tag them on such objects, not only standalone clocks, and this might be making visual noise.

**Comment by james2432** [comment]
Aparently they are never tagged with churches/towers:
https://taginfo.openstreetmap.org/tags/amenity=clock#combinations

**Comment by kocio-pl** [comment]
I meant "on the wall of the churches/towers", not "the churches/towers which are considered to be clocks" - this wouldn't make sense for me.

**Comment by james2432** [comment]
do you have an example of a clock that is near a church?

**Comment by kocio-pl** [comment]
No, I don't.

**Comment by lakedistrictOSM** [comment]
Try https://www.openstreetmap.org/way/350214201#map=19/52.79738/-0.03797, https://www.openstreetmap.org/way/474490616 and https://www.openstreetmap.org/way/144553454

**Comment by james2432** [comment]
https://www.openstreetmap.org/#map=19/51.31811/9.49671
Z18:
image

Z19:
image

https://www.openstreetmap.org/way/350214201#map=17/52.79765/-0.03705
Z18:
image

Z19:
image

https://www.openstreetmap.org/way/214357782#map=19/35.78913/-78.82871
Z18:
image

Z19:
image

**Comment by Tomasz-W** [comment]
I'm considering which tag should have priority, but in cases which I can imagine at the moment:
-amenity=clock > man_made=tower
-amenity=clock < man_made=tower + tower:type=bell_tower

**Comment by kocio-pl** [comment]
I would exclude at least `support=wall` ("The feature is supported by a wall.") and `support=wall_mounted` ("The feature is mounted on a wall."). Feel free to suggest something more, which indicates it's not a standalone clock or just a part of something more important (see https://wiki.openstreetmap.org/wiki/Key:support ). The example:

https://www.openstreetmap.org/node/2470273778

**Comment by HolgerJeromin** [comment]
> I would exclude at least support=wall

This could lead to dropping this tag to get it rendered....

**Comment by kocio-pl** [comment]
@Tomasz-W The problem of objects with multiple tags is hard, basically I just rely on current code. Otherwise we should have complicated tools for chopping and moving some queries.

@HolgerJeromin Cheating is always possible, but there should be a way for a proper tagging to not make a visual mess in typical cases.

**Comment by kocio-pl** [comment]
@james2432 How are your plans about updating this code?

**Comment by james2432** [comment]
i'm waiting for a consensus to be made on what are the filters...

**Comment by kocio-pl** [comment]
I'm not afraid of potential tag dropping - I don't believe this could be real problem, but even if it was, it would be just less complete, but still accurate tagging.

**Comment by kocio-pl** [comment]
Are you still waiting? It doesn't look to me like it will be more clear than it already is, we just have different opinions.

**Comment by Tomasz-W** [comment]
In case if @james2432 have time and will to extend the code, I'll remind my picks of filters (it's not so nessesary, but I think it may be useful to avoid not-expected icons on map)

amenity=clock > man_made=tower
amenity=clock < man_made=tower + tower:type=bell_tower
amenity=clock < advertising=column

**Comment by james2432** [comment]
sorry, i've been busy dealing with a failing hard-drive. With that summary I should be able to get it up and running by this weekend

**Comment by james2432** [comment]
From:https://www.openstreetmap.org/#map=19/35.78917/-78.82883
z19:
image
z18:
image

From: https://www.openstreetmap.org/#map=19/51.31811/9.49671
z19:
image

z18:
image

From: https://www.openstreetmap.org/way/350214201#map=17/52.79765/-0.03705
z18:
image
z19:
image

Fixed the query. The priority over bell tower was a bit complicated

File: docker-compose.yml [hunk 1]
```diff
@@ -19,6 +19,8 @@ services:
     build:
       context: .
       dockerfile: Dockerfile.db
+    ports:
+      - "5432:5432"
```
Comment (kocio-pl) [inline docker-compose.yml]: This code doesn't belong here.

Comment (james2432) [reply docker-compose.yml]: oops. Fixed

**Comment by Tomasz-W** [comment]
@james2432 I don't like bold outline in Osmic icon, but I understand that 4 o'clock may be more readable, so I changed my icon a little bit and uploaded it Gist: https://gist.github.com/Tomasz-W/07ed36017474669db82b11cb7f64dd26

Can you test it to compare them?

**Comment by meased** [comment]
Another option would be to fill the face of the clock and "cut out" the hands. This would make the icon stronger, but eliminate some noise between the outline and the hands:

clock
https://gist.github.com/meased/3c8ec416fc1d4fe0e5e295d49eb7d1d0

Also, pixel alignment on the hands is slightly off.

**Comment by james2432** [comment]
Looks better:
Z19:
image

**Comment by Tomasz-W** [comment]
@kocio-pl Is it ready to merge or there is something more needed?

**Comment by james2432** [comment]
I'd really like to have things merged before I start on other things (say the craft schema) as it just involves more work for me to fix up the project.mml for every pull request

**Comment by kocio-pl** [comment]
Sorry for lack of response:
1. I still would exclude support=wall and support=wall_mounted. 
2. The code should have parentheses like in #3245

**Comment by james2432** [comment]
is it just when checking IS NULL or any time we are checking accessors?

**Comment by kocio-pl** [comment]
I don't know, @Zverik, could you explain it?

**Comment by james2432** [comment]
From what I'm reading it could have to do specifically NULL results.

The () would dereference the value from being from an hstore(i think there were issues in older versions of postgres doing any action with a null hstore)

**Comment by Zverik** [comment]
`tags->'tower:type'` should be in parentheses, but the more pressing question is, why CI hasn't caught up on it.

**Comment by Zverik** [comment]
Ah, indeed. Could somebody please merge #3252 so this type of errors doesn't happen?

**Comment by james2432** [comment]
Proof of exclusion for support = wall and support=wall_mounted
https://www.openstreetmap.org/way/458324854  (way, support=wall_mounted) (z19 ... z20 was also tested)
image

https://www.openstreetmap.org/node/5507892508 (node, support=wall_mounted) (z19 ... z20 was also tested)
image

https://www.openstreetmap.org/way/442511850 (way, support=wall) (z19 ... z20 was also tested)
image

https://www.openstreetmap.org/node/4294148104 (node, support=wall) (z19 ... z20 was also tested)
image

Also added the tags->  in parenthesis when using IS

File: project.mml [hunk 2]
```diff
@@ -1519,6 +1519,8 @@ Layer:
             tags->'tower:type' as "tower:type",
             tags->'castle_type' as castle_type,
             tags->'information' as information,
+            tags->'visibility' as "visibility",
+            tags->'support' as support,
```
Comment (kocio-pl) [inline project.mml]: Why "visibility" is in quotes and support is not?

File: project.mml [hunk 3]
```diff
@@ -1652,6 +1654,8 @@ Layer:
...
```
Comment (kocio-pl) [inline project.mml]: The same as above.

File: project.mml [hunk 4]
```diff
@@ -2343,6 +2349,7 @@ Layer:
           WHERE building IS NOT NULL
             AND building NOT IN ('no')
             AND name IS NOT NULL
+            AND NOT tags->'tower:type' IN('bell_tower')
```
Comment (kocio-pl) [inline project.mml]: Lacking space.

File: project.mml (hunk 2 above)
Comment (james2432) [reply project.mml]: purly esthetical, will add the quotes to follow others

**Comment by james2432** [comment]
https://github.com/james2432/openstreetmap-carto/blob/d0d9723662e8baaa5662e6091f172c7cea4dd8dd/project.mml#L2295

See purely esthetical, the other tags/column names are not enclosed with quotes here.

Added space.

Is there a coding standard that says all column names must be enclosed in quotes?

File: project.mml [hunk 5]
```diff
@@ -2343,6 +2349,7 @@ Layer:
           WHERE building IS NOT NULL
             AND building NOT IN ('no')
             AND name IS NOT NULL
+            AND NOT tags->'tower:type' IN ('bell_tower')
```
Comment (Zverik) [inline project.mml]: Better `NOT IN`, as above with `building NOT IN ('no')`.

**Comment by kocio-pl** [comment]
I might have few days off without the access to the computer, so I will look at it soon after I'm back.

**Comment by kocio-pl** [comment]
Sorry, wrong button...

File: project.mml (hunk 5 above)
Comment (james2432) [reply project.mml]: Changed

**Comment by kocio-pl** [comment]
This code is strange, it has two sections:

```CSS
[feature = 'amenity_clock']["tower:type" != 'bell_tower'][feature != 'advertising_column'][zoom >= 18]
```

```CSS
[feature = 'amenity_clock']["tower:type" != 'bell_tower'][feature != 'advertising_column']['support' != 'wall']['support' != 'wall_mounted'][zoom >= 19] { 
```

Wall clocks should never be shown, because they are just a part of tower (unlike clock towers with multiple faces for example), not the landmarks in themselves.

**Comment by Tomasz-W** [comment]
@james2432 What's the state of this PR? Are you willing to finish it?

**Comment by james2432** [comment]
well my laptop just died completely(motherboard), so I will be unable to continue this until I get a new one.

**Comment by Tomasz-W** [comment]
Oh, I understand. Anyway, I still hope that someday we will have these clocks and craft objects on map :)

**Comment by Adamant36** [comment]
@kocio-pl, if I want to finish this do I do it by opening a new PR and noting that it closes this one along with the original issue or is there a certain procedure for that?

**Comment by kocio-pl** [comment]
New PR with closing comment will be OK.

**Comment by james2432** [comment]
you can create a PR  on my branch I will merge it, ai'm just unable to create new changes since I dont have a computer to work on it(using phone)

**Comment by Adamant36** [comment]
OK cool. Creating a new PR here instead of on a branch might make it easier to deal with any subsequent issues that could come up.

**Comment by Adamant36** [comment]
@james2432, I copied everything except the wall thing from your commit and although Kosmtik ran, I couldn't get it to display clocks on the map. If there anyway you could go over my code and see what I might be keeping them from displaying? https://github.com/Adamant36/openstreetmap-carto/tree/clock

**Comment by Tomasz-W** [comment]
@Adamant36 Oh, one more thing. Please change icon to "v1" version as it's obviously more clear than a "v2"   

https://gist.github.com/Tomasz-W/07ed36017474669db82b11cb7f64dd26

**Comment by matkoniecz** [comment]
I am really, really sorry but I think that showing all clocks is not a good idea (in other words I think that underlying #3070 should be rejected).

**Comment by Adamant36** [comment]
I thought it was a little wierd myself, but it seemed like it had some support at the time. 

Its generally easy to go back to an issue after some time has passed, or the disscussion has died down, and to close it because it doesn't make sense in retrospect. You might not get any complaints either. Since everyone has pretty much left the building at this point. It still might be worth leaving open due to the earlier support though. 

I was still planning on working on it eventual if james2434 never came back. 

What specific issue do you have with it? Maybe some clocks could be displayed instead of all clocks or something.

**Comment by Adamant36** [comment]
@tomasz-W if you want to open a new issue for the display of prominent ones as @matkoniecz suggests in the original issue I'd be willing to work on it when I get some time if I can. That would at least allow for some discussion of the pros and cons in the meantime though.

**Pull request closed without merging**